File: src/main/java/rbq/wtf/lycoris/client/wrapper/wrappers/wrapper/util/Vec3i.java
package rbq.wtf.lycoris.client.wrapper.wrappers.wrapper.util;

import rbq.wtf.lycoris.client.wrapper.MapEnum;
import rbq.wtf.lycoris.client.wrapper.wrappers.annotation.WrapConstructor;
import rbq.wtf.lycoris.client.wrapper.wrappers.annotation.WrapField;
import rbq.wtf.lycoris.client.wrapper.wrappers.annotation.WrapMethod;
import rbq.wtf.lycoris.client.wrapper.wrappers.annotation.WrapperClass;
import rbq.wtf.lycoris.client.wrapper.wrappers.utils.ReflectUtil;
import rbq.wtf.lycoris.client.wrapper.wrappers.wrapper.IWrapper;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;


@WrapperClass(mcpName = "net.minecraft.util.Vec3i", targetMap = MapEnum.VANILLA189)
public class Vec3i extends IWrapper {
    @WrapField(mcpName = "x", targetMap = MapEnum.VANILLA189)
    public static Field x;
    @WrapField(mcpName = "y", targetMap = MapEnum.VANILLA189)
    public static Field y;
    @WrapField(mcpName = "z", targetMap = MapEnum.VANILLA189)
    public static Field z;
    @WrapConstructor(targetMap = MapEnum.VANILLA189, signature = {double.class, double.class, double.class})
    public static Constructor Vec3i_DDD;
    @WrapMethod(mcpName = "distanceSq", targetMap = MapEnum.VANILLA1122, signature = "(DDD)D")
    public static Method distanceSq;

    public Vec3i(Object obj) {
        super(obj);
    }

    public Vec3i(double xIn, double yIn, double zIn) {
        this(ReflectUtil.construction(Vec3i_DDD, xIn, yIn, zIn));
    }

    public int getX() {
        return (int) ReflectUtil.getField(x, getWrapObject());
    }

    public int getY() {
        return (int) ReflectUtil.getField(y, getWrapObject());
    }

    public int getZ() {
        return (int) ReflectUtil.getField(z, getWrapObject());
    }

    public double distanceTo(double posX, double posY, double posZ) {
        return (double) invoke(distanceSq, posX, posY, posZ);
    }
}

File: src/main/java/rbq/wtf/lycoris/client/wrapper/wrappers/util/Vec3i.kt
package rbq.wtf.lycoris.client.wrapper.wrappers.util

import rbq.wtf.lycoris.client.wrapper.MapEnum
import rbq.wtf.lycoris.client.wrapper.annotation.WrapConstructor
import rbq.wtf.lycoris.client.wrapper.annotation.WrapField
import rbq.wtf.lycoris.client.wrapper.annotation.WrapMethod
import rbq.wtf.lycoris.client.wrapper.annotation.WrapperClass
import rbq.wtf.lycoris.client.wrapper.utils.ReflectUtil
import rbq.wtf.lycoris.client.wrapper.IWrapper
import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.Method

@WrapperClass(mcpName = "net.minecraft.util.Vec3i", targetMap = MapEnum.VANILLA189)
open class Vec3i(obj: Any) : IWrapper(obj) {
    constructor(xIn: Double, yIn: Double, zIn: Double) : this(ReflectUtil.construction(Vec3i_DDD, xIn, yIn, zIn))

    companion object {
        @WrapField(mcpName = "x", targetMap = MapEnum.VANILLA189)
        lateinit var x: Field

        @WrapField(mcpName = "y", targetMap = MapEnum.VANILLA189)
        lateinit var y: Field

        @WrapField(mcpName = "z", targetMap = MapEnum.VANILLA189)
        lateinit var z: Field

        @WrapConstructor(targetMap = MapEnum.VANILLA189, signature = [Double::class, Double::class, Double::class])
        lateinit var Vec3i_DDD: Constructor<*>

        @WrapMethod(mcpName = "distanceSq", targetMap = MapEnum.VANILLA1122, signature = "(DDD)D")
        lateinit var distanceSq: Method
    }
    
    val x: Int
        get() = getField(Companion.x) as Int
    val y: Int
        get() = getField(Companion.y) as Int
    val z: Int
        get() = getField(Companion.z) as Int

    fun distanceTo(posX: Double, posY: Double, posZ: Double): Double {
        return invoke(distanceSq, posX, posY, posZ) as Double
    }
}